File: prototype/php/src/app/Domain/Cart/CartTotals.php
<?php

namespace App\Domain\Cart;

use App\Domain\Money\Money;
use DomainException;

final readonly class CartTotals
{
    /**
     * @param  array<int, Money>  $subtotalsBySeller
     */
    private function __construct(
        public int $itemCount,
        public Money $subtotal,
        private array $subtotalsBySeller,
    ) {}

    /**
     * @param  list<CartLine>  $lines
     */
    public static function forCheckout(array $lines): self
    {
        if ($lines === []) {
            throw new DomainException('An order needs at least one item.');
        }

        return self::from($lines);
    }

    /**
     * @param  list<CartLine>  $lines
     */
    public static function from(array $lines): self
    {
        $itemCount = 0;
        $subtotal = Money::fromCents(0);
        $subtotalsBySeller = [];

        foreach ($lines as $line) {
            $itemCount += $line->quantity;
            $subtotal = $subtotal->add($line->total());
            $sellerSubtotal = $subtotalsBySeller[$line->sellerId] ?? Money::fromCents(0);
            $subtotalsBySeller[$line->sellerId] = $sellerSubtotal->add($line->total());
        }

        ksort($subtotalsBySeller);

        return new self($itemCount, $subtotal, $subtotalsBySeller);
    }

    /**
     * @return array<int, Money>
     */
    public function subtotalsBySeller(): array
    {
        return $this->subtotalsBySeller;
    }
}

File: prototype/rails/src/app/domain/cart/cart_totals.rb
require_relative "../money"
require_relative "cart_line"

module Domain
  module Cart
    # What a cart is worth, whole and split by the seller each line belongs to.
    CartTotals = Data.define(:item_count, :subtotal, :subtotals_by_seller) do
      def self.from(lines)
        new(
          item_count: lines.sum(&:quantity),
          subtotal: total_of(lines),
          subtotals_by_seller: lines.group_by(&:seller_id).transform_values { |own| total_of(own) }.sort.to_h.freeze
        )
      end

      def self.for_checkout(lines)
        raise ArgumentError, "an order needs at least one item" if lines.empty?

        from(lines)
      end

      def self.total_of(lines)
        lines.sum(Money.from_cents(0), &:total)
      end
      private_class_method :total_of
    end
  end
end
